cpp/src/table/row_operators.cu

```diff
@@ -393,6 +393,83 @@ namespace lexicographic {
+ * new ones. This is because the new child was generated by ranking and always has zero
+ * offset thus cannot replace the input child if it is sliced.
+ *
+ * The new generated offsets column needs to be returned and kept alive.
+ *
+ * @param[in] input The input column_view of type LIST
+ * @param[in] new_child A new child column to replace the existing child of the input
+ * @param[out] out_cols An array to store the new generated offsets (if applicable)
+ * @param[in] stream CUDA stream used for device memory operations and kernel launches
+ * @return An output column_view with child replaced
+ */
+auto replace_child(column_view const& input,
+                   column_view const& new_child,
+                   std::vector<std::unique_ptr<column>>& out_cols,
+                   rmm::cuda_stream_view stream)
+{
+  auto const make_output = [&input](auto const& offsets_cv, auto const& child_cv) {
+    return column_view{data_type{type_id::LIST},
+                       input.size(),
+                       nullptr,
+                       input.null_mask(),
+                       input.null_count(),
+                       0,
+                       {offsets_cv, child_cv}};
+  };
+
+  if (input.offset() == 0) {
+    return make_output(input.child(lists_column_view::offsets_column_index), new_child);
+  }
+
+  out_cols.emplace_back(cudf::lists::detail::get_normalized_offsets(
+    lists_column_view{input}, stream, rmm::mr::get_current_device_resource()));
+  return make_output(out_cols.back()->view(), new_child);
+}
+
+/**
+ * @brief Compute ranks of the input column.
+ *
+ * Dense rank type is used for compute ranking of the input for later lexicographic comparison.
+ *
+ * Consider this example: `input = [ [{0, "a"}, {3, "c"}], [{0, "a"}, {2, "b"}] ]`.
+ * If first rank is used, `transformed_input = [ [0, 3], [1, 2] ]`. Comparing them will lead
+ * to the result row(0) < row(1) which is incorrect.
+ * With dense rank, `transformed_input = [ [0, 2], [0, 1] ]`, producing the correct output for
+ * lexicographic comparison.
+ *
+ * In addition, since the being ranked input column is always a nested child column instead of
+ * top-level column, the column order for ranking should be fixed to the same value
+ * `order::ASCENDING` in all situations.
+ * For example, with the same input above, using column order as `order::ASCENDING` we will have
+ * `transformed_input = [ [0, 2], [0, 1] ]`. The output of sorting `transformed_input` will be
+ * exactly the same as sorting `input` regardless of the sorting order (ASC or DESC).
+ *
+ * @param input The input column to compute ranks
+ * @param column_null_order The flag indicating how nulls compare to non-null values
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return The output rank columns
+ */
+auto compute_ranks(column_view const& input,
+                   null_order column_null_order,
+                   rmm::cuda_stream_view stream)
```

Comment:
The functions changed in this file (and perhaps others not changed here) should take an `mr` that is used for the memory they allocate. This is consistent with the [developer guide section on output memory](https://github.com/rapidsai/cudf/blob/branch-23.06/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#output-memory), which includes `detail` APIs.